File: packages/settings/ios/Sources/SettingsPlugin/SettingsConfig.swift
import Foundation
import Capacitor

/**
 Plugin configuration container.

 This struct is responsible for reading and exposing static configuration
 values defined under the `Settings` key in `capacitor.config.ts`.

 Configuration is:
 - read once during plugin initialization
 - treated as immutable runtime input
 - consumed only by native code
 */
public struct SettingsConfig {

    // MARK: - Configuration Keys

    // Define configuration keys for consistency
    private struct Keys {
        static let verboseLogging = "verboseLogging"
    }

    // MARK: - Public Config Values

    /**
     Enables verbose native logging.

     When enabled, additional debug information is printed
     to the Xcode console via the plugin logger.

     Default: false
     */
    public let verboseLogging: Bool

    // MARK: - Private Defaults

    // Default values
    private static let defaultVerboseLogging = false

    // MARK: - Init

    /**
     Initializes the configuration by reading values from the Capacitor bridge.

     - Parameter plugin: The CAPPlugin instance used to access typed configuration.
     */
    init(plugin: CAPPlugin) {
        let config = plugin.getConfig()

        // Bool
        verboseLogging =
            config.getBoolean(Keys.verboseLogging, Self.defaultVerboseLogging)
    }
}

File: packages/settings/android/src/main/java/io/capkit/settings/config/SettingsConfig.kt
package io.capkit.settings.config

import android.content.Context
import com.getcapacitor.Plugin
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Data Transfer Object (DTO) for parsing capacitor.config.ts configuration.
 */
@Serializable
private data class SettingsConfigData(
  @SerialName("verboseLogging")
  val verboseLogging: Boolean = false,
)

/**
 * Plugin configuration holder for the Settings plugin.
 *
 * This class is responsible for reading and parsing static configuration values
 * defined under the `plugins.Settings` key in `capacitor.config.ts`.
 *
 * Configuration is read once during plugin initialization and treated as
 * immutable runtime input.
 */
class SettingsConfig(
  plugin: Plugin,
) {
  /**
   * Android application context.
   * Exposed for native components that may require it.
   */
  val context: Context = plugin.context

  /**
   * Enables verbose / debug logging for the plugin.
   *
   * When enabled, additional logs are printed to Logcat via [Logger.debug].
   *
   * Default: false
   */
  val verboseLogging: Boolean

  init {
    verboseLogging = parseConfig(plugin).verboseLogging
  }

  private companion object {
    private val json =
      Json {
        ignoreUnknownKeys = true
        isLenient = true
      }

    private fun parseConfig(plugin: Plugin): SettingsConfigData =
      try {
        val configJsonObject = plugin.config.getConfigJSON()
        if (configJsonObject != null) {
          json.decodeFromString<SettingsConfigData>(configJsonObject.toString())
        } else {
          SettingsConfigData()
        }
      } catch (_: Exception) {
        SettingsConfigData()
      }
  }
}
